File: ImperatorToCK3/Source/Imperator/Genes/AccessoryGenes.cpp
#include "AccessoryGenes.h"
#include "Log.h"
#include "ParserHelpers.h"
#include "CommonRegexes.h"



Imperator::AccessoryGenes::AccessoryGenes(std::istream& theStream) {
	registerKeys();
	parseStream(theStream);
	clearRegisteredKeywords();
}


void Imperator::AccessoryGenes::registerKeys() {
	registerKeyword("index", [this](std::istream& theStream) {
		index = commonItems::getInt(theStream);
	});
	registerMatcher(commonItems::stringMatch, [this](const std::string& geneName, std::istream& theStream) {
		genes.emplace(geneName, AccessoryGene(theStream));
	});
	registerMatcher(commonItems::catchallRegexMatch, commonItems::ignoreItem);
}

File: ImperatorToCK3/Imperator/Genes/AccessoryGenes.cs
using System.Collections.Generic;
using commonItems;

namespace ImperatorToCK3.Imperator.Genes {
	public class AccessoryGenes : Parser {
		public uint Index { get; private set; } = 0;
		public Dictionary<string, AccessoryGene> Genes { get; private set; } = new();

		public AccessoryGenes() { }
		public AccessoryGenes(BufferedReader reader) {
			RegisterKeys();
			ParseStream(reader);
			ClearRegisteredRules();
		}
		private void RegisterKeys() {
			RegisterKeyword("index", reader => {
				Index = (uint)new SingleInt(reader).Int;
			});
			RegisterRegex(CommonRegexes.StringRegex, (reader, geneName) => {
				Genes.Add(geneName, new AccessoryGene(reader));
			});
			RegisterRegex(CommonRegexes.Catchall, ParserHelpers.IgnoreItem);
		}
	}
}
